File: src/services/__tests__/simulationService.test.ts
/**
 * simulationService 테스트
 *
 * GWT 형식으로 테스트 구조화:
 * - Given (사전 조건): Mock 설정 및 테스트 데이터 준비
 * - When (동작): 서비스 메서드 호출
 * - Then (결과): API 호출 검증 및 반환값 확인
 */

import axios from "@/lib/axios";
import { simulationService } from "../simulationService";

// axios 모듈 모킹
jest.mock("@/lib/axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe("simulationService", () => {
  beforeEach(() => {
    // 각 테스트 전 mock 초기화
    jest.clearAllMocks();
  });

  /**
   * 시뮬레이션 생성 API 테스트
   */
  describe("create", () => {
    it("시뮬레이션 생성 요청을 올바르게 보내야 한다", async () => {
      // Given: 시뮬레이션 생성 성공 응답이 설정된 상태
      const mockResponse = { data: { id: 1, message: "Created successfully" } };
      mockedAxios.post.mockResolvedValueOnce(mockResponse);

      // When: create 메서드를 호출하면
      const result = await simulationService.create("005930", "2024-01-01");

      // Then: 쿼리 파라미터가 포함된 엔드포인트로 요청하고 결과를 반환해야 함
      expect(mockedAxios.post).toHaveBeenCalledWith("/api/simulation?code=005930&start=2024-01-01");
      expect(result).toEqual({ id: 1, message: "Created successfully" });
    });
  });

  /**
   * 시뮬레이션 목록 조회 API 테스트
   */
  describe("getList", () => {
    it("시뮬레이션 목록을 조회해야 한다", async () => {
      // Given: 시뮬레이션 목록 응답이 설정된 상태
      const mockSimulations = {
        data: {
          simulations: [
            {
              id: 1,
              stockCode: "005930",
              startDate: "2024-01-01",
              endDate: null,
              createdAt: "2024-01-01T00:00:00Z",
            },
          ],
          totalElements: 1,
        },
      };
      mockedAxios.get.mockResolvedValueOnce(mockSimulations);

      // When: getList 메서드를 호출하면
      const result = await simulationService.getList();

      // Then: 시뮬레이션 목록과 총 개수를 반환해야 함
      expect(mockedAxios.get).toHaveBeenCalledWith("/api/simulation");
      expect(result.simulations).toHaveLength(1);
      expect(result.totalElements).toBe(1);
    });
  });

  /**
   * 시뮬레이션 종료 API 테스트
   */
  describe("terminate", () => {
    it("시뮬레이션 종료 요청을 올바르게 보내야 한다", async () => {
      // Given: 시뮬레이션 종료 성공 응답이 설정된 상태
      const mockResponse = { data: { message: "Terminated successfully" } };
      mockedAxios.patch.mockResolvedValueOnce(mockResponse);

      // When: terminate 메서드를 호출하면
      const result = await simulationService.terminate(1, "2024-12-31");

      // Then: 종료 날짜가 포함된 엔드포인트로 PATCH 요청해야 함
      expect(mockedAxios.patch).toHaveBeenCalledWith("/api/simulation/1?end=2024-12-31");
      expect(result).toEqual({ message: "Terminated successfully" });
    });
  });

  /**
   * 시뮬레이션 삭제 API 테스트
   */
  describe("delete", () => {
    it("시뮬레이션 삭제 요청을 올바르게 보내야 한다", async () => {
      // Given: 시뮬레이션 삭제 성공 응답이 설정된 상태
      mockedAxios.delete.mockResolvedValueOnce({});

      // When: delete 메서드를 호출하면
      await simulationService.delete(1);

      // Then: 시뮬레이션 ID로 DELETE 요청해야 함
      expect(mockedAxios.delete).toHaveBeenCalledWith("/api/simulation/1");
    });
  });

  /**
   * 거래 데이터 추가 API 테스트
   */
  describe("addTrade", () => {
    it("거래 데이터 추가 요청을 올바르게 보내야 한다", async () => {
      // Given: 거래 데이터와 성공 응답이 설정된 상태
      const tradeData = {
        timestamp: "2024-01-15",
        balance: 0,
        price: 70000,
        upper: 80000,
        lower: 60000,
        type: "B" as const,
        volume: 10,
        comment: "Test buy",
      };
      const mockResponse = { data: { message: "Trade added successfully" } };
      mockedAxios.post.mockResolvedValueOnce(mockResponse);

      // When: addTrade 메서드를 호출하면
      const result = await simulationService.addTrade(1, tradeData);

      // Then: 시뮬레이션 ID의 data 엔드포인트로 거래 데이터를 POST해야 함
      expect(mockedAxios.post).toHaveBeenCalledWith("/api/simulation/1/data", tradeData);
      expect(result).toEqual({ message: "Trade added successfully" });
    });
  });

  /**
   * 거래 내역 조회 API 테스트
   */
  describe("getTrades", () => {
    it("거래 내역을 조회해야 한다", async () => {
      // Given: 거래 내역 응답이 설정된 상태
      const mockTrades = {
        data: {
          trades: [
            {
              timestamp: "2024-01-15",
              balance: 0,
              price: 70000,
              upper: 80000,
              lower: 60000,
              type: "B" as const,
              volume: 10,
            },
          ],
          totalElements: 1,
        },
      };
      mockedAxios.get.mockResolvedValueOnce(mockTrades);

      // When: getTrades 메서드를 호출하면
      const result = await simulationService.getTrades(1);

      // Then: 시뮬레이션 ID의 data 엔드포인트에서 거래 내역을 반환해야 함
      expect(mockedAxios.get).toHaveBeenCalledWith("/api/simulation/1/data");
      expect(result.trades).toHaveLength(1);
      expect(result.totalElements).toBe(1);
    });
  });

  /**
   * 시뮬레이션 리포트 생성 API 테스트
   */
  describe("generateReport", () => {
    it("시뮬레이션 리포트 생성을 요청해야 한다", async () => {
      // Given: 시뮬레이션 리포트 생성 성공 응답이 설정된 상태
      const mockResponse = { data: { message: "Report generated successfully" } };
      mockedAxios.post.mockResolvedValueOnce(mockResponse);

      // When: generateReport 메서드를 호출하면
      const result = await simulationService.generateReport(1);

      // Then: 시뮬레이션 리포트 생성 요청을 보내고 메시지를 반환해야 함
      expect(mockedAxios.post).toHaveBeenCalledWith("/api/simulation/1/report");
      expect(result).toEqual({ message: "Report generated successfully" });
    });
  });

  /**
   * 시뮬레이션 리포트 조회 API 테스트
   */
  describe("getReport", () => {
    it("시뮬레이션 리포트를 조회해야 한다", async () => {
      // Given: 시뮬레이션 리포트 응답이 설정된 상태
      const mockReport = {
        data: {
          summary: {
            simulationId: 1,
            stockCode: "005930",
            startDate: "2024-01-01",
            endDate: "2024-12-31",
            finalAvgPrice: 75000,
            totalInvestment: 7000000,
            totalRealizedProfit: 500000,
            totalRoi: 7.14,
            totalScore: 85,
          },
          trades: [],
          totalSellVolume: 100,
          totalTradeCount: 10,
        },
      };
      mockedAxios.get.mockResolvedValueOnce(mockReport);

      // When: getReport 메서드를 호출하면
      const result = await simulationService.getReport(1);

      // Then: 시뮬레이션 리포트 데이터를 반환해야 함
      expect(mockedAxios.get).toHaveBeenCalledWith("/api/simulation/1/report");
      expect(result.summary.simulationId).toBe(1);
      expect(result.summary.totalRoi).toBe(7.14);
    });
  });
});

File: src/test/java/aib/trademate/domain/simulation/service/SimulationServiceTest.java
package aib.trademate.domain.simulation.service;

import aib.trademate.domain.member.entity.Member;
import aib.trademate.domain.member.repository.MemberRepository;
import aib.trademate.domain.simulation.dto.*;
import aib.trademate.domain.simulation.entity.Simulation;
import aib.trademate.domain.simulation.entity.SimulationTrade;
import aib.trademate.domain.simulation.entity.TradeType;
import aib.trademate.domain.simulation.repository.SimulationReportRepository;
import aib.trademate.domain.simulation.repository.SimulationRepository;
import aib.trademate.domain.simulation.repository.SimulationTradeRepository;
import aib.trademate.global.exception.BusinessException;
import aib.trademate.global.exception.ErrorCode;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.util.ReflectionTestUtils;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.*;

/**
 * SimulationService 단위 테스트
 * 
 * 시뮬레이션 CRUD 및 거래 관리 로직을 검증합니다.
 */
@ExtendWith(MockitoExtension.class)
@DisplayName("SimulationService 테스트")
@SuppressWarnings("null")
class SimulationServiceTest {

    @Mock
    private SimulationRepository simulationRepository;

    @Mock
    private SimulationTradeRepository tradeRepository;

    @Mock
    private SimulationReportRepository reportRepository;

    @Mock
    private MemberRepository memberRepository;

    @Mock
    private SimulationScoreCalculator scoreCalculator;

    @InjectMocks
    private SimulationService simulationService;

    private Member testMember;
    private Simulation testSimulation;

    @BeforeEach
    void setUp() {
        testMember = createMember(1L, "test@test.com");
        testSimulation = createSimulation(1L, testMember.getId(), "005930", LocalDate.of(2024, 1, 1));
    }

    @Nested
    @DisplayName("시뮬레이션 생성 테스트")
    class CreateSimulationTest {

        @Test
        @DisplayName("시뮬레이션 생성이 성공한다")
        void createSimulationSuccess() {
            // given
            String email = "test@test.com";
            String stockCode = "005930";
            LocalDate startDate = LocalDate.of(2024, 1, 1);

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.save(any(Simulation.class))).willAnswer(invocation -> {
                Simulation saved = invocation.getArgument(0);
                ReflectionTestUtils.setField(saved, "id", 1L);
                return saved;
            });

            // when
            CreateSimulationResponse response = simulationService.createSimulation(email, stockCode, startDate);

            // then
            assertThat(response.id()).isEqualTo(1L);
            verify(simulationRepository).save(any(Simulation.class));
        }

        @Test
        @DisplayName("존재하지 않는 회원으로 시뮬레이션 생성 시 예외가 발생한다")
        void createSimulationFailsWhenMemberNotFound() {
            // given
            String email = "notfound@test.com";
            given(memberRepository.findByEmail(email)).willReturn(Optional.empty());

            // when & then
            assertThatThrownBy(() -> 
                    simulationService.createSimulation(email, "005930", LocalDate.now()))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.RESOURCE_NOT_FOUND);
                    });
        }
    }

    @Nested
    @DisplayName("시뮬레이션 종료일 업데이트 테스트")
    class UpdateEndDateTest {

        @Test
        @DisplayName("종료일 업데이트가 성공한다")
        void updateEndDateSuccess() {
            // given
            String email = "test@test.com";
            LocalDate endDate = LocalDate.of(2024, 3, 1);

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when
            simulationService.updateEndDate(email, testSimulation.getId(), endDate);

            // then
            assertThat(testSimulation.getEndDate()).isEqualTo(endDate);
        }

        @Test
        @DisplayName("다른 사용자의 시뮬레이션 종료일 업데이트 시 예외가 발생한다")
        void updateEndDateFailsWhenNotOwner() {
            // given
            Member otherMember = createMember(2L, "other@test.com");
            String email = "other@test.com";

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(otherMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when & then
            assertThatThrownBy(() -> 
                    simulationService.updateEndDate(email, testSimulation.getId(), LocalDate.now()))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.ACCESS_DENIED);
                    });
        }
    }

    @Nested
    @DisplayName("시뮬레이션 목록 조회 테스트")
    class GetMySimulationsTest {

        @Test
        @DisplayName("내 시뮬레이션 목록을 조회한다")
        void getMySimulationsSuccess() {
            // given
            String email = "test@test.com";
            List<Simulation> simulations = List.of(
                    testSimulation,
                    createSimulation(2L, testMember.getId(), "035720", LocalDate.of(2024, 2, 1))
            );

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findByMemberId(testMember.getId())).willReturn(simulations);

            // when
            SimulationListResponse response = simulationService.getMySimulations(email);

            // then
            assertThat(response.simulations()).hasSize(2);
            assertThat(response.simulations().get(0).stockCode()).isEqualTo("005930");
            assertThat(response.simulations().get(1).stockCode()).isEqualTo("035720");
        }

        @Test
        @DisplayName("시뮬레이션이 없으면 빈 목록을 반환한다")
        void getMySimulationsReturnsEmptyList() {
            // given
            String email = "test@test.com";
            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findByMemberId(testMember.getId())).willReturn(List.of());

            // when
            SimulationListResponse response = simulationService.getMySimulations(email);

            // then
            assertThat(response.simulations()).isEmpty();
        }
    }

    @Nested
    @DisplayName("시뮬레이션 삭제 테스트")
    class DeleteSimulationTest {

        @Test
        @DisplayName("시뮬레이션 삭제가 성공한다")
        void deleteSimulationSuccess() {
            // given
            String email = "test@test.com";

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when
            simulationService.deleteSimulation(email, testSimulation.getId());

            // then
            verify(simulationRepository).delete(testSimulation);
        }

        @Test
        @DisplayName("존재하지 않는 시뮬레이션 삭제 시 예외가 발생한다")
        void deleteSimulationFailsWhenNotFound() {
            // given
            String email = "test@test.com";
            Long nonExistentId = 999L;

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(nonExistentId)).willReturn(Optional.empty());

            // when & then
            assertThatThrownBy(() -> 
                    simulationService.deleteSimulation(email, nonExistentId))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.SIMULATION_NOT_FOUND);
                    });
        }
    }

    @Nested
    @DisplayName("거래 추가 테스트")
    class AddTradeTest {

        @Test
        @DisplayName("거래 추가가 성공한다")
        void addTradeSuccess() {
            // given
            String email = "test@test.com";
            AddTradeRequest request = new AddTradeRequest(
                    LocalDate.of(2024, 1, 15),
                    1000,     // balance (Integer)
                    70000,    // price (Integer)
                    75000,    // upper (Integer)
                    65000,    // lower (Integer)
                    "B",      // 매수
                    10,       // volume
                    "첫 매수"
            );

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when
            simulationService.addTrade(email, testSimulation.getId(), request);

            // then
            verify(tradeRepository).save(any(SimulationTrade.class));
        }

        @Test
        @DisplayName("잘못된 거래 타입으로 거래 추가 시 예외가 발생한다")
        void addTradeFailsWithInvalidTradeType() {
            // given
            String email = "test@test.com";
            AddTradeRequest request = new AddTradeRequest(
                    LocalDate.of(2024, 1, 15),
                    1000,
                    70000,
                    75000,
                    65000,
                    "X",  // Invalid type
                    10,
                    "잘못된 타입"
            );

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when & then
            assertThatThrownBy(() -> 
                    simulationService.addTrade(email, testSimulation.getId(), request))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.INVALID_INPUT_VALUE);
                    });
        }
    }

    @Nested
    @DisplayName("거래 목록 조회 테스트")
    class GetTradesTest {

        @Test
        @DisplayName("거래 목록 조회가 성공한다")
        void getTradesSuccess() {
            // given
            String email = "test@test.com";
            List<SimulationTrade> trades = List.of(
                    createTrade(1L, testSimulation, TradeType.BUY, 10),
                    createTrade(2L, testSimulation, TradeType.SELL, 5)
            );

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));
            given(tradeRepository.findBySimulationIdOrderByTradeDateAsc(testSimulation.getId())).willReturn(trades);

            // when
            TradeListResponse response = simulationService.getTrades(email, testSimulation.getId());

            // then
            assertThat(response.trades()).hasSize(2);
        }
    }

    @Nested
    @DisplayName("보고서 생성 테스트")
    class GenerateReportTest {

        @Test
        @DisplayName("종료된 시뮬레이션의 보고서를 생성한다")
        void generateReportSuccess() {
            // given
            String email = "test@test.com";
            testSimulation.updateEndDate(LocalDate.of(2024, 3, 1));

            List<SimulationTrade> trades = List.of(
                    createTrade(1L, testSimulation, TradeType.BUY, 10),
                    createTrade(2L, testSimulation, TradeType.SELL, 10)
            );

            SimulationReportResponse expectedReport = new SimulationReportResponse(
                    ReportSummary.builder()
                            .simulationId(testSimulation.getId())
                            .stockCode("005930")
                            .startDate(LocalDate.of(2024, 1, 1))
                            .endDate(LocalDate.of(2024, 3, 1))
                            .finalAvgPrice(BigDecimal.valueOf(70000))
                            .totalInvestment(BigDecimal.valueOf(700000))
                            .totalRealizedProfit(BigDecimal.valueOf(50000))
                            .totalRoi(BigDecimal.valueOf(7.14))
                            .totalScore(BigDecimal.valueOf(85))
                            .build(),
                    List.of(),
                    10,
                    2
            );

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));
            given(reportRepository.existsBySimulationId(testSimulation.getId())).willReturn(false);
            given(tradeRepository.findBySimulationIdOrderByTradeDateAsc(testSimulation.getId())).willReturn(trades);
            given(scoreCalculator.calculateReport(testSimulation, trades)).willReturn(expectedReport);

            // when
            simulationService.generateReport(email, testSimulation.getId());

            // then
            verify(scoreCalculator).calculateReport(testSimulation, trades);
            verify(reportRepository).save(any());
        }

        @Test
        @DisplayName("종료되지 않은 시뮬레이션의 보고서 생성 시 예외가 발생한다")
        void generateReportFailsWhenNotEnded() {
            // given
            String email = "test@test.com";
            // endDate is null (not ended)

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));

            // when & then
            assertThatThrownBy(() ->
                    simulationService.generateReport(email, testSimulation.getId()))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.SIMULATION_NOT_ENDED);
                    });

            verify(scoreCalculator, never()).calculateReport(any(), any());
        }

        @Test
        @DisplayName("이미 보고서가 존재하는 시뮬레이션에 보고서 생성 시 409 예외가 발생한다")
        void generateReportFailsWhenAlreadyExists() {
            // given
            String email = "test@test.com";
            testSimulation.updateEndDate(LocalDate.of(2024, 3, 1));

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));
            given(reportRepository.existsBySimulationId(testSimulation.getId())).willReturn(true);

            // when & then
            assertThatThrownBy(() ->
                    simulationService.generateReport(email, testSimulation.getId()))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.REPORT_ALREADY_EXISTS);
                    });

            verify(scoreCalculator, never()).calculateReport(any(), any());
        }
    }

    @Nested
    @DisplayName("시뮬레이션 보고서 조회 테스트")
    class GetReportTest {

        @Test
        @DisplayName("저장된 보고서를 조회한다")
        void getReportSuccess() {
            // given
            String email = "test@test.com";
            testSimulation.updateEndDate(LocalDate.of(2024, 3, 1));

            aib.trademate.domain.simulation.entity.SimulationReport savedReport =
                    aib.trademate.domain.simulation.entity.SimulationReport.builder()
                            .simulationId(testSimulation.getId())
                            .finalAvgPrice(BigDecimal.valueOf(70000))
                            .totalInvestment(BigDecimal.valueOf(700000))
                            .totalRealizedProfit(BigDecimal.valueOf(50000))
                            .totalRoi(BigDecimal.valueOf(7.14))
                            .totalScore(BigDecimal.valueOf(85))
                            .totalSellVolume(10)
                            .totalTradeCount(2)
                            .build();

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));
            given(reportRepository.findBySimulationId(testSimulation.getId())).willReturn(Optional.of(savedReport));

            // when
            SimulationReportResponse response = simulationService.getReport(email, testSimulation.getId());

            // then
            assertThat(response.summary().totalScore()).isEqualTo(BigDecimal.valueOf(85));
            assertThat(response.summary().stockCode()).isEqualTo("005930");
            verify(reportRepository).findBySimulationId(testSimulation.getId());
            verify(scoreCalculator, never()).calculateReport(any(), any());
        }

        @Test
        @DisplayName("보고서가 존재하지 않으면 404 예외가 발생한다")
        void getReportFailsWhenNotFound() {
            // given
            String email = "test@test.com";

            given(memberRepository.findByEmail(email)).willReturn(Optional.of(testMember));
            given(simulationRepository.findById(testSimulation.getId())).willReturn(Optional.of(testSimulation));
            given(reportRepository.findBySimulationId(testSimulation.getId())).willReturn(Optional.empty());

            // when & then
            assertThatThrownBy(() ->
                    simulationService.getReport(email, testSimulation.getId()))
                    .isInstanceOf(BusinessException.class)
                    .satisfies(ex -> {
                        BusinessException be = (BusinessException) ex;
                        assertThat(be.getErrorCode()).isEqualTo(ErrorCode.REPORT_NOT_FOUND);
                    });
        }
    }

    // ========== Helper Methods ==========

    private Member createMember(Long id, String email) {
        Member member = Member.builder()
                .email(email)
                .password("encodedPassword")
                .name("테스터")
                .birthdate(LocalDate.of(1990, 1, 1))
                .phone("010-1234-5678")
                .build();
        ReflectionTestUtils.setField(member, "id", id);
        return member;
    }

    private Simulation createSimulation(Long id, Long memberId, String stockCode, LocalDate startDate) {
        Simulation simulation = Simulation.builder()
                .memberId(memberId)
                .stockCode(stockCode)
                .startDate(startDate)
                .build();
        ReflectionTestUtils.setField(simulation, "id", id);
        return simulation;
    }

    private SimulationTrade createTrade(Long id, Simulation simulation, TradeType type, int volume) {
        SimulationTrade trade = SimulationTrade.builder()
                .simulation(simulation)
                .tradeDate(LocalDate.of(2024, 1, 15))
                .balance(1000)
                .price(70000)
                .upperLimit(75000)
                .lowerLimit(65000)
                .tradeType(type)
                .volume(volume)
                .comment("테스트 거래")
                .build();
        ReflectionTestUtils.setField(trade, "id", id);
        return trade;
    }
}
